src/Nethermind/Nethermind.Peering.Mothra/MothraLibp2p.cs

```diff
@@ -0,0 +1,240 @@
+//  it under the terms of the GNU Lesser General Public License as published by
+//  the Free Software Foundation, either version 3 of the License, or
+//  (at your option) any later version.
+// 
+//  The Nethermind library is distributed in the hope that it will be useful,
+//  but WITHOUT ANY WARRANTY; without even the implied warranty of
+//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+//  GNU Lesser General Public License for more details.
+// 
+//  You should have received a copy of the GNU Lesser General Public License
+//  along with the Nethermind. If not, see <http://www.gnu.org/licenses/>.
+
+using System;
+using System.Collections.Generic;
+using System.Runtime.InteropServices;
+using System.Text;
+using Microsoft.Extensions.Logging;
+
+namespace Nethermind.Peering.Mothra
+{
+    public class MothraLibp2p : IMothraLibp2p
+    {
+        private readonly Dictionary<LogLevel, string> _debugLevels = new Dictionary<LogLevel, string>()
+        {
+            [LogLevel.Trace] = "trace",
+            [LogLevel.Debug] = "debug",
+            [LogLevel.Information] = "info",
+            [LogLevel.Warning] = "warn",
+            [LogLevel.Error] = "error",
+            [LogLevel.Critical] = "crit"
+        };
+
+        private GCHandle _argsHandle;
+        private readonly MothraInterop.DiscoveredPeer _discoveredPeer;
+        private GCHandle _discoveredPeerHandle;
+        private readonly MothraInterop.ReceiveGossip _receiveGossip;
+        private GCHandle _receiveGossipHandle;
+        private readonly MothraInterop.ReceiveRpc _receiveRpc;
+        private GCHandle _receiveRpcHandle;
+
+        public MothraLibp2p()
+        {
+            unsafe
+            {
+                _discoveredPeer = DiscoveredPeerHandler;
+                _receiveGossip = ReceiveGossipHandler;
+                _receiveRpc = ReceiveRpcHandler;
+            }
+
+            _discoveredPeerHandle = GCHandle.Alloc(_discoveredPeer);
+            _receiveGossipHandle = GCHandle.Alloc(_receiveGossip);
+            _receiveRpcHandle = GCHandle.Alloc(_receiveRpc);
+        }
+
+        public event EventHandler<GossipReceivedEventArgs>? GossipReceived;
+
+        public event EventHandler<PeerDiscoveredEventArgs>? PeerDiscovered;
+
+        public event EventHandler<RpcReceivedEventArgs>? RpcReceived;
+
+        public void SendGossip(string topic, ReadOnlySpan<byte> data)
+        {
+            byte[] topicUtf8 = Encoding.UTF8.GetBytes(topic);
+            unsafe
+            {
+                fixed (byte* topicUtf8Ptr = topicUtf8)
+                fixed (byte* dataPtr = data)
+                {
+                    MothraInterop.SendGossip(topicUtf8Ptr, topicUtf8.Length, dataPtr, data.Length);
+                }
+            }
+        }
+
+        public void SendRpcRequest(string method, string peer, ReadOnlySpan<byte> data)
+        {
+            byte[] methodUtf8 = Encoding.UTF8.GetBytes(method);
```

Comment:
Maybe some buffer pool / array pool here - this will be most likely the most allocation heavy part of the app.

Reply:
Or, since the methods will be same, we should cache the methods as byte arrays somehwere. Also for each peer we should cache its byte[] format not to allocate over and over again.